colossalai/kernel/triton/flash_decoding.py

```diff
@@ -181,97 +182,127 @@ def _flash_decoding_fwd_reduce_kernel(
-    context_lengths: torch.Tensor,  # [batch_size]
-    block_tables: torch.Tensor,  # [batch_size, max_blocks_per_sequence]
+def flash_decoding_attention(
+    q: torch.Tensor,
+    k_cache: torch.Tensor,
+    v_cache: torch.Tensor,
+    kv_seq_len: torch.Tensor,
+    block_tables: torch.Tensor,
     block_size: int,
-    num_kv_group: int = 1,
+    max_seq_len_in_batch: int = None,
+    mid_output: torch.Tensor = None,
+    mid_output_lse: torch.Tensor = None,
+    sm_scale: int = None,
+    kv_group_num: int = 1,
 ):
-    bsz, _, num_heads, head_dim = q.shape
+    """
+    Flash decoding implemented with a blocked KV Cache (PagedAttention) during decoding stage.
+
+    Args:
+        q (torch.Tensor):       [bsz, num_heads, q_len(1), head_dim]
+        k_cache (torch.Tensor): [num_blocks, num_kv_heads, head_dim, block_size]
+        v_cache (torch.Tensor): [num_blocks, num_kv_heads, head_dim, block_size]
+        kv_seq_len (torch.Tensor): [batch_size]
+        block_tables (torch.Tensor): [batch_size, max_blocks_per_sequence]
+        max_seq_len_in_batch (int): Maximum sequence length in the batch.
+        mid_output (torch.Tensor): [ max_bsz , num_heads, kv_max_split_num, head_dim]
+            Intermediate output tensor. `max_bsz` should be greater than or equal to `bsz`.
+        mid_output_lse (torch.Tensor): [ max_bsz , num_heads, kv_max_split_num]
+            Log-sum-exp of intermediate output. `max_bsz` should be greater than or equal to `bsz`.
+        block_size (int): Size of each block in the blocked key/value cache.
+        num_kv_group (int, optional): Number of key/value groups. Defaults to 1.
+
+    Returns:
+        Output tensor with shape [bsz, num_heads, q_len, head_dim]
+    """
+    bsz, num_heads, _, head_dim = q.shape
 
     assert head_dim in {32, 64, 128, 256}
-    assert context_lengths.shape[0] == block_tables.shape[0] == bsz, (
+    assert kv_seq_len.shape[0] == block_tables.shape[0] == bsz, (
         f"Got incompatible batch size (number of seqs):\n"
-        f"  Conext lengths bsz {context_lengths.shape[0]}, Block tables bsz {block_tables.shape[0]}, "
+        f"  KV seq lengths bsz {kv_seq_len.shape[0]}, Block tables bsz {block_tables.shape[0]}, "
         f"batch size {bsz}"
     )
     assert k_cache.size(-1) == v_cache.size(-1) == block_size, (
         f"Got incompatible block size on kv caches:\n"
         f"  assigned block_size {block_size}, k_cache block_size {k_cache.size(-1)}, "
         f"v_cache block_size {v_cache.size(-1)}"
     )
-    # NOTE `context_lengths` records the (kv) sequence lengths incorporating past kv sequence lengths.
-    bsz = context_lengths.size(0)  # e.g. the number of seqs
-    max_seq_len = context_lengths.max().item()
-    sm_scale = 1.0 / (head_dim**0.5)
+    # NOTE `kv_seq_len` records the (kv) sequence lengths incorporating past kv sequence lengths.
+    bsz = kv_seq_len.size(0)  # e.g. the number of seqs
```

Comment:
The batch size has been obtained in line 218, and the code here is redundant.

Reply:
revised